design/libraries/nodes-and-kinds.md

```diff
@@ -0,0 +1,152 @@
+Nodes and Kinds
+===============
+
+Your IPLD library should have two key types:
+
+1. `Node`;
+2. `Kind`.
+
+`Node` should be an interface -- the membership should be open
+(aka, it should be possible for other packages to implement it).
+
+`Kind` should be an enumeration -- a fixed set of named members,
+which should not be extendable.
+
+
+Kind
+----
+
+`Kind` maps very directly onto the definition of
+[Data Model Kinds](../../data-model-layer/data-model.md#kinds).
+
+`Kind` does not include the Schema layer's concept of "struct", etc.
+
+`Kind` must be an enum, **and not a sum type**.  Attempting to implement
+kind as a sum type conflates it with `Node`.
+(This may be tempting to try to combine `Kind` and `Node` into a single
+sum type definition if you're only looking at the Data Model layer,
+but it is a mistake: both Schema types and Advanced Layouts require
+the ability to add more implementations of `Node`, so this conflation
+will cause cataclysmic problems and force a painful refactor
+as soon as you get to implementing those systems.
+See the [different implementors of Node](#different-implementors-of-node)
+section, later in this document, for more information on this.)
+
+
+Node
+----
+
+`Node` is a monomorphized interface for handling data -- in other words,
+we make all data look and act like a `Node`, so that we can write all of our
+functions against the `Node` interface, and have that work for any sort of data.
+
+`Node` has functions for examining any of the
+[Data Model Kinds](../../data-model-layer/data-model.md#kinds).
+For example, this means `Node` must be able to
+do a key lookup for a map kind,
```

Comment:
linebreaking pattern is getting pretty weird here, in most of the doc it kind of looks like you're breaking close to 80 chars on the nearest punctuation, but it breaks down in this para.

Reply:
Yeah, I blithely follow my heart.  I try to aim for "eighty-_ish_", but heavy on the _-ish_.  I'll also readily break on commas or sentences.

My overall rule is "try to make the diff minimally irritating in future rounds", so the punctuation heuristic helps with that, whereas a character count enforced strictly is irritation-maximalizing.

I don't think we'll get the equivalent of `go fmt` for natural language any time soon, though, so I try not to lose too much sleep over anything.